spyro/sources/Sources.py

```diff
@@ -19,75 +21,40 @@ def __init__(self, model, mesh, V, comm):
-            Contains a list of excitations
+        Receivers: :class: 'Receiver' object
 
         """
 
-        self.source_type = model["acquisition"]["source_type"]
-        self.num_sources = model["acquisition"]["num_sources"]
-        self.pos = model["acquisition"]["source_pos"]
-        self.dimension = model["opts"]["dimension"]
         self.mesh = mesh
-        self.V = V
-        self.my_ensemble = comm
-
-        @property
-        def num_sources(self):
-            return self.__num_sources
-
-        @num_sources.setter
-        def num_sources(self, value):
-            assert value > 0, "Number of sources must be > 0"
-            self.__num_sources = value
-
-        @property
-        def pos(self):
-            return self.__pos
-
-        @pos.setter
-        def pos(self, value):
-            self.__pos = value
-
-    def create(self):
-        """ Create injection operator(s) to excite source(s)."""
-
-        V = self.V
-
-        if self.dimension == 2:
-            z, x = SpatialCoordinate(self.mesh)
-        elif self.dimension == 3:
-            z, x, y = SpatialCoordinate(self.mesh)
-
-        if self.dimension == 2:
-            source = Constant([0, 0])
-            if self.source_type == "Ricker":
-                delta = Interpolator(delta_expr(source, z, x), V)
-            elif self.source_type == "MMS":
-                delta = Interpolator(MMS_space(source, z, x), V)
-            else:
-                raise ValueError("Unrecognized source type")
-        elif self.dimension == 3:
-            source = Constant([0, 0, 0])
-            if self.source_type == "Ricker":
-                delta = Interpolator(delta_expr_3d(source, z, x, y), V)
-            elif self.source_type == "MMS":
-                delta = Interpolator(MMS_space_3d(source, z, x, y), V)
-        else:
-            raise ValueError("Incorrect dimension")
-
-        excitations = []
-        for x0 in self.pos:
-            source.assign(x0)
-            excitations.append(Function(delta.interpolate()))
-
-        return excitations
+        self.space = V
+        self.my_ensemble = my_ensemble
+        self.dimension = model["opts"]["dimension"]
+        self.degree = model["opts"]["degree"]
+
+        self.num_receivers = model["acquisition"]["num_sources"]
+        self.receiver_locations = model["acquisition"]["source_pos"]
+        self.source_type = model["acquisition"]["source_type"]
+
+        self.cellIDs = None
+        self.cellVertices = None
+        self.cell_tabulations = None
+        self.cellNodeMaps = None
+        self.nodes_per_cell = None
+
+    def apply_source(self, rhs_forcing,value):
+        """ Applies source in a assembled right hand side.
+        """
+        for source_id in range(self.num_receivers):
+            for i in range(len(self.cellNodeMaps[source_id])):
+                rhs_forcing.dat.data[int(self.cellNodeMaps[source_id][i])] = value * self.cell_tabulations[source_id][i]
```

Comment:
Does this assume only one source?

Reply:
No, it goes through every source with for source_id in range(self.num_receivers). It says num_receivers because itś a subclass from Receivers, but that value comes from the number of sources in the input file in the __init__.

Reply:
It should work with multiple sources and parallelism since receivers works with multiple receivers and parallelism. However this still needs to be tested to make sure.

This PR was done in the testing_eage branch so I could compare it with the figures we generated for EAGE Latin AMerica to see if there really is an improvement.  I am still running the grid sweep to compare and that's going to take awhile. After that, if it really is better, I am going to test multiple sources and parallelism before doing a PR on the master branch